data_structures/src/chain.rs

```diff
@@ -784,6 +882,236 @@ pub enum InventoryItem {
+        let info = DataRequestInfo::default();
+        let stage = DataRequestStage::COMMIT;
+
+        Self {
+            data_request,
+            info,
+            stage,
+            epoch,
+        }
+    }
+
+    /// Add commit
+    pub fn add_commit(&mut self, output_pointer: OutputPointer) {
+        assert_eq!(self.stage, DataRequestStage::COMMIT);
+        self.info.commits.insert(output_pointer);
+    }
+
+    /// Add reveal
+    pub fn add_reveal(&mut self, output_pointer: OutputPointer) {
+        assert_eq!(self.stage, DataRequestStage::REVEAL);
+        self.info.reveals.insert(output_pointer);
+    }
+
+    /// Add tally and return the data request report
+    pub fn add_tally(
+        mut self,
+        output_pointer: OutputPointer,
+    ) -> (DataRequestOutput, DataRequestReport) {
+        assert_eq!(self.stage, DataRequestStage::TALLY);
+        self.info.tally = Some(output_pointer);
+        // This try_from can only fail if the tally is None, and we have just set it to Some
+        (
+            self.data_request,
+            DataRequestReport::try_from(self.info).unwrap(),
+        )
+    }
+
+    /// Advance to the next stage, returning true on success.
+    /// Since the data requests are updated by looking at the transactions from a valid block,
+    /// the only issue would be that there were no commits in that block.
+    pub fn update_stage(&mut self) -> bool {
+        let old_stage = self.stage;
+
+        self.stage = match self.stage {
+            DataRequestStage::COMMIT => {
+                if self.info.commits.is_empty() {
+                    DataRequestStage::COMMIT
+                } else {
+                    DataRequestStage::REVEAL
+                }
+            }
+            DataRequestStage::REVEAL => {
+                if self.info.reveals.is_empty() {
+                    DataRequestStage::REVEAL
+                } else {
+                    DataRequestStage::TALLY
+                }
+            }
+            DataRequestStage::TALLY => {
+                if self.info.tally.is_none() {
+                    DataRequestStage::TALLY
+                } else {
+                    panic!("Data request in tally stage should have been removed from the pool");
+                }
+            }
+        };
+
+        self.stage != old_stage
+    }
+}
+
+/// Data request current stage
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub enum DataRequestStage {
+    /// Expecting commitments for data request
+    COMMIT,
+    /// Expecting reveals to previously published commitments
+    REVEAL,
+    /// Expecting tally to be included in block
+    TALLY,
+}
+
+/// Pool of active data requests
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct ActiveDataRequestPool {
+    /// Current active data request, in which this node has announced commitments.
+    /// Key: Data Request Pointer, Value: Reveal Transaction
```

Comment:
This comment does not provide any useful information.

Reply:
I left it as it clarifies what are we storing in the `HashMap`. :)